File: Top Interview 150/Hard/N-QueensII.cpp
#include <unordered_set>
#include <vector>
#include <iostream>
#include <chrono>
#include <numeric>

class Solution{
    int res;
    int n;
    std::vector<int> cols;
    std::vector<int> neg_diagonal;
    std::vector<int> pos_diagonal;
    std::vector<int> unassigned_cols;

public:
    int totalNQueens(int x) {
        res = 0;
        n = x;
        cols.resize(n);
        neg_diagonal.resize(n * 2);
        pos_diagonal.resize(n * 2);
        unassigned_cols.resize(n);
        std::iota(unassigned_cols.begin(), unassigned_cols.end(), 0);
        
        backtrack(0);
        return res;
    }

private:
    void backtrack(int row) {
        if (row == n) {
            res++;
            return;
        }
        
        std::vector<int> unassigned_copy = unassigned_cols;
        for (int col : unassigned_copy) {
            int curr_neg_diag = row - col + n;
            int curr_pos_diag = row + col;
            
            if (neg_diagonal[curr_neg_diag] || pos_diagonal[curr_pos_diag]) 
                continue;

            cols[col]++;
            neg_diagonal[curr_neg_diag]++;
            pos_diagonal[curr_pos_diag]++;
            unassigned_cols.erase(std::find(unassigned_cols.begin(), unassigned_cols.end(), col));

            backtrack(row + 1);
            
            cols[col]--;
            neg_diagonal[curr_neg_diag]--;
            pos_diagonal[curr_pos_diag]--;
            unassigned_cols.push_back(col);
        }
    }
};

int main () {
    std::cout << "Hello World\n";

    Solution sol = Solution();
    
    std::vector<std::tuple<int, int>> tests =  {
        {1, 1},
        {2, 0},
        {3, 0},
        {4, 2}, 
        {5, 10},
        {6, 4},
        {7, 40},
        {8, 92},
        {9, 352},
        {15, 2279184},
    };
    
    std::vector<bool> all_tests;
    std::vector<std::tuple<int, int>> res;
    std::vector<std::tuple<int, int, int>> failed;
    bool ok = true;
    
    auto start = std::chrono::high_resolution_clock::now();
    int i = 1;
    for (const auto& test : tests) {
        int n, expected;
        std::tie(n, expected) = test;
        int output = sol.totalNQueens(n);
        res.push_back({output, expected});
        if (output != expected) {
            ok = false;
            failed.push_back({i, output, expected});
            all_tests.push_back(false);
            i++;
            continue;
        }
        all_tests.push_back(true);
        i++;
    }

    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    std::cout << "This run took " << duration.count() << " microseconds." << std::endl;
    
    if (ok) {
        std::cout << std::endl;
        std::cout << "\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/ Passed All Tests \\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/\\/" << std::endl;
        std::cout << std::endl;
    }
    else {
        std::cout << "---------------------------------- Failed Tests ----------------------------------" << std::endl;
        for (const auto& fail : failed) {
            int test_num, output, expected;
            std::tie(test_num, output, expected) = fail;
            std::cout << "Test Number: " << test_num << std::endl;
            std::cout << "    Output:   " << output << std::endl;
            std::cout << "    Expected: " << expected << std::endl;
        }
        std::cout << std::endl;
        std::cout << "----------------------------------- All Tests ------------------------------------" << std::endl;
        i = 0;
        for (const auto& r : res) {
            int o, e;
            std::tie(o, e) = r;
            std::cout << "Test Number: " << i + 1;
            if (all_tests[i]) std::cout << " Pass" << std::endl;
            else std::cout << " Failed" << std::endl;
            std::cout << "    Output:   " << o << std::endl;
            std::cout << "    Expected: " << e << std::endl;
            std::cout << std::endl;
            i++;
        }
    }
}

File: Top Interview 150/Hard/N-QueensII.py
import time

class Solution:
    def totalNQueens(self, n: int) -> int:
        def backtrack(row:int) -> None:
            nonlocal res
            if row == n:
                res += 1
                return

            for col in unassigned_cols.copy():
                curr_neg_diag = row - col + n
                curr_pos_diag = row + col
                
                if neg_diagonal[curr_neg_diag] or pos_diagonal[curr_pos_diag]:
                    continue

                cols[col] = True
                neg_diagonal[curr_neg_diag] = True
                pos_diagonal[curr_pos_diag] = True
                unassigned_cols.remove(col)

                backtrack(row + 1)
                
                cols[col] = False
                neg_diagonal[curr_neg_diag] = False
                pos_diagonal[curr_pos_diag] = False
                unassigned_cols.append(col)

        res = 0
        cols, neg_diagonal, pos_diagonal = [False] * n, [False] * (n * 2), [False] * (n * 2)
        unassigned_cols = list(i for i in range(n))
        backtrack(0)
        return res

if __name__ == "__main__":
    sol = Solution()
    tests =  [
                (1, 1),
                (2, 0),
                (3, 0),
                (4, 2), 
                (5, 10),
                (6, 4),
                (7, 40),
                (8, 92),
                (9, 352),
                # (15, 2279184),
            ]
    
    all_tests = []
    res = []
    failed = []
    ok = True
    
    start_time = time.time()
    for i, test in enumerate(tests, start=1):
        n, expected = test
        output = sol.totalNQueens(n)
        res.append((output, expected))
        if output != expected:
            ok = False
            failed.append((i, output, expected))
            all_tests.append(False)
            continue
        all_tests.append(True)

    end_time = time.time()
    elapsed_time = end_time - start_time
    print(f"Runtime: {round(elapsed_time, 9)}s")
    
    if ok:
        print()
        print("\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/ Passed All Tests \/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/")
        print()
    else:
        print("---------------------------------- Failed Tests ----------------------------------")
        for fail in failed:
            test_num, output, expected = fail
            print(f"Test Number: {test_num}") 
            print(f"    Output:   {output}")
            print(f"    Expected: {expected}")
        print()
        print("----------------------------------- All Tests ------------------------------------")
        for i, (o, e) in enumerate(res):
                print(f"Test Number: {i+1}    ", end="") 
                print("Pass") if all_tests[i] else print("Failed")
                print(f"    Output:   {o}")
                print(f"    Expected: {e}")
                print()